### PR TITLE
SLING-8392 - Create sub-command to manage the Jira update when promoting a release

### PR DESCRIPTION
**Pull request by raducotescu**

### CODE REVIEW

File: src/main/java/org/apache/sling/cli/impl/jira/Version.java [hunk 1]
```diff
@@ -16,10 +16,13 @@
  */
 package org.apache.sling.cli.impl.jira;
 
+@SuppressWarnings("unused")
```
Comment (rombert) [inline src/main/java/org/apache/sling/cli/impl/jira/Version.java]: Is this needed?

File: src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java [hunk 2]
```diff
@@ -160,35 +178,102 @@ public void create(String versionName) throws IOException {
...
     public List<Issue> findFixedIssues(Release release) throws IOException {
-        return findIssues("is not empty", release);
+        return findIssues(release).stream().filter(issue -> issue.getResolution() != null).collect(Collectors.toList());
     }
 
-    private List<Issue> findIssues(String resolution, Release release) throws IOException {
+    private void closeIssues(List<Issue> issues) throws Exception {
+        List<Promise<Issue>> closedIssues = new ArrayList<>();
+        for (Issue issue : issues) {
+            if (!"Closed".equals(issue.getStatus())) {
```
Comment (rombert) [inline src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java]: A constant would be nice

File: src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java [hunk 3]
```diff
@@ -160,35 +178,102 @@ public void create(String versionName) throws IOException {
...
+            }
+        }
+        Promise<List<Issue>> closedFixedIssues = promiseFactory.all(closedIssues);
+        Throwable failed = closedFixedIssues.getFailure();
+        if (failed != null) {
+            if (failed instanceof FailedPromisesException) {
+                FailedPromisesException failedPromisesException = (FailedPromisesException) failed;
+                StringBuilder failureMessages = new StringBuilder();
+                for (Promise<?> promise : failedPromisesException.getFailedPromises()) {
+                    failureMessages.append(promise.getFailure().getMessage()).append("\n");
```
Comment (rombert) [inline src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java]: Could we somehow preserve the stack traces?

File: src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java [hunk 4]
```diff
@@ -160,35 +178,102 @@ public void create(String versionName) throws IOException {
...
+            } else {
+                throw new Exception(failed);
+            }
+        }
+    }
+
+    public void release(Release release) throws Exception {
+        List<Issue> issues = findIssues(release);
+        List<Issue> unresolvedIssues = new ArrayList<>();
+        issues.forEach(issue -> {
```
Comment (rombert) [inline src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java]: filtering first would be idiomatic

File: src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java [hunk 5]
```diff
@@ -341,4 +420,54 @@ private void moveIssueToNewVersion(Version oldVersion, Version newVersion, Issue
             throw new RuntimeException(e);
         }
     }
+
+    private Promise<Transition> getCloseTransition(Issue issue) {
+        HttpGet get = newGet("issue/" + issue.getId() + "/transitions");
+        try {
```
Comment (rombert) [inline src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java]: You could collapse the `catch` block in the nested `try`

File: src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java [hunk 6]
```diff
@@ -341,4 +420,54 @@ private void moveIssueToNewVersion(Version oldVersion, Version newVersion, Issue
...
+            try (CloseableHttpClient client = httpClientFactory.newClient()) {
+                try (CloseableHttpResponse getResponse = client.execute(get, httpClientFactory.newPreemptiveAuthenticationContext())) {
+                    try (InputStream getContent = getResponse.getEntity().getContent();
+                         InputStreamReader getReader = new InputStreamReader(getContent)) {
+                        if (getResponse.getStatusLine().getStatusCode() != 200) {
+                            throw newException(getResponse, getReader);
+                        }
+                        Gson gson = new Gson();
+                        List<Transition> transitions = gson.fromJson(getReader, TransitionsResponse.class).getTransitions();
+                        Optional<Transition> transition = transitions.stream().filter(t -> "Close Issue".equals(t.getName())).findFirst();
```
Comment (rombert) [inline src/main/java/org/apache/sling/cli/impl/jira/VersionClient.java]: Again, a constant would be nice.

File: src/test/java/org/apache/sling/cli/impl/jira/EditVersionJiraAction.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            }
+
+        } catch (NumberFormatException e) {
+            LOGGER.error("Unable to parse version id from " + ex.getRequestURI().getPath(), e);
+            ex.sendResponseHeaders(400, -1);
+        }
+        return true;
+    }
+
+    @SuppressWarnings("unused")
```
Comment (rombert) [inline src/test/java/org/apache/sling/cli/impl/jira/EditVersionJiraAction.java]: Why is this needed?

**Comment by raducotescu** [comment]
Thanks! I'll merge and then address your comments. The `@SupressWarnings` is my slight OCD kicking in when looking at the warnings my IDE generates. To be honest, we should refactor all our Gson "POJOs" to provide only fields and getters since Gson anyways only cares about the fields for serialization [0].

[0] - https://github.com/google/gson/blob/master/GsonDesignDocument.md - see the **Using fields vs getters to indicate Json elements** section.
